src/txn/txn.c

```diff
@@ -1025,10 +1029,26 @@ __txn_fixup_prepared_update(
                     WT_STAT_CONN_INCR(
                       session, txn_prepare_rollback_fix_hs_update_with_ckpt_reserved_txnid);
                 }
+
+                /* The update can be inserted to the history store again. */
+                F_CLR(fix_upd, WT_UPDATE_TO_DELETE_FROM_HS | WT_UPDATE_HS);
             } else
                 WT_STAT_CONN_INCR(session, txn_prepare_rollback_do_not_remove_hs_update);
-        } else
+        } else {
             WT_ERR(hs_cursor->remove(hs_cursor));
+            /* The update can be inserted to the history store again. */
+            F_CLR(fix_upd, WT_UPDATE_TO_DELETE_FROM_HS | WT_UPDATE_HS);
+            if (fix_upd->type == WT_UPDATE_TOMBSTONE) {
+                /* We may not find a full update following the tombstone if it is obsolete. */
+                for (fix_upd = fix_upd->next; fix_upd != NULL; fix_upd = fix_upd->next)
+                    if (fix_upd->txnid != WT_TXN_ABORTED)
+                        break;
+                WT_ASSERT(session,
+                  fix_upd != NULL && F_ISSET(fix_upd, WT_UPDATE_RESTORED_FROM_HS | WT_UPDATE_HS));
```

Comment:
It is a restored update and tombstone, there is no way that there can be another aborted update in the middle. Mention in a comment that it is possible only when the update exists in the memory.